Improve misleading ORMInvalidArgumentException message

jonny-no1 opened the pull request:
See https://github.com/doctrine/doctrine2/issues/6029#issuecomment-340304346.

Commit:
Add failing test for misleading exception message

Describes: https://github.com/doctrine/doctrine2/issues/6029

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * @ManyToMany(targetEntity="DDC6029User", mappedBy="groups")
+     */
+    private $users;
```
Is this needed? Let's keep only the things required by the tests

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    public function testIssue()
+    {
+        $user = new DDC6029User();
```
Shouldn't we add here the expected exception and the expected exception message? Like:

```php
$this->expectException(ORMInvalidArgumentException::class);
$this->expectExceptionMessage('Expected value of type "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group" for association field "Doctrine\Tests\ORM\Functional\Ticket\DDC6029User#$groups", got "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group2" instead.');
```

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php:
```diff
@@ -0,0 +1,83 @@
+     * Verifies that when wrong entity is persisted via relationship field, the error message does not correctly state
+     * the expected class name.
+     */
+    public function testIssue()
```
` : void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php:
```diff
@@ -0,0 +1,83 @@
+
+    /**
+     * Verifies that when wrong entity is persisted via relationship field, the error message does not correctly state
+     * the expected class name.
```
Please add `@group` with the ticket number so that we can trace things back

Commit:
Use the correct type for the exception message

Since the UoW checks each item of a *-to-many association to ensure
it has the correct type, we should never say that we expect an instance
of `Doctrine\Common\Collections\Collection` or an `array`.

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php:
```diff
@@ -0,0 +1,79 @@
+        $user = new DDC6029User();
+        $user->addGroup(new DDC6029Group2());
+
+        self::expectException(ORMInvalidArgumentException::class);
```
This method is not static, so we should use `$this` instead

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6029Test.php:
```diff
@@ -0,0 +1,79 @@
+        $user->addGroup(new DDC6029Group2());
+
+        self::expectException(ORMInvalidArgumentException::class);
+        self::expectExceptionMessage('Expected value of type "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group" for association field "Doctrine\Tests\ORM\Functional\Ticket\DDC6029User#$groups", got "Doctrine\Tests\ORM\Functional\Ticket\DDC6029Group2" instead.');
```
This method is not static, so we should use `$this` instead

lcobucci commented on lib/Doctrine/ORM/ORMInvalidArgumentException.php:
```diff
@@ -195,11 +195,7 @@ public static function invalidIdentifierBindingEntity()
     {
-        $expectedType = 'Doctrine\Common\Collections\Collection|array';
-
-        if (($assoc['type'] & ClassMetadata::TO_ONE) > 0) {
```
Is there any explanation on why we had this code there (commit msg)? Maybe @Ocramius can give us some context

foaly-nr1 replied:
No details in the commit message of https://github.com/doctrine/doctrine2/commit/71a6a88de82d23232cdf5377a44804a5f55d29b8 and not sure where I could find details on DDC-3490.

lcobucci reviewed (approved):
I've rebased and added a test for one-to-many association as well.

lcobucci commented:
@foaly-nr1 🚢 thanks for your contribution!
